File: py/entity/get_postal_codes_by_city_entity.py
# ZippopotamusZipCode SDK GetPostalCodesByCity entity

from __future__ import annotations

from utility.voxgig_struct import voxgig_struct as vs
from core import helpers
from zippopotamuszipcode_types import (
    GetPostalCodesByCity,
    GetPostalCodesByCityListMatch,
)


class GetPostalCodesByCityEntity:

    def __init__(self, client, entopts=None):
        if entopts is None:
            entopts = {}
        if "active" not in entopts:
            entopts["active"] = True
        elif entopts["active"] is False:
            pass  # keep false
        else:
            entopts["active"] = True

        self._name = "get_postal_codes_by_city"
        self._client = client
        self._utility = client.get_utility()
        self._entopts = entopts
        self._data = {}
        self._match = {}

        self._entctx = self._utility.make_context({
            "entity": self,
            "entopts": entopts,
        }, client.get_root_ctx())

        self._utility.feature_hook(self._entctx, "PostConstructEntity")

    def get_name(self):
        return self._name

    def make(self):
        opts = {}
        for k, v in self._entopts.items():
            opts[k] = v
        return GetPostalCodesByCityEntity(self._client, opts)

    def data_set(self, args=None):
        if args is not None:
            self._data = helpers.to_map(vs.clone(args)) or {}
            self._utility.feature_hook(self._entctx, "SetData")

    def data_get(self) -> GetPostalCodesByCity:
        self._utility.feature_hook(self._entctx, "GetData")
        return vs.clone(self._data)

    def match_set(self, args=None):
        if args is not None:
            self._match = helpers.to_map(vs.clone(args)) or {}
            self._utility.feature_hook(self._entctx, "SetMatch")

    def match_get(self) -> GetPostalCodesByCity:
        self._utility.feature_hook(self._entctx, "GetMatch")
        return vs.clone(self._match)

    def stream(self, action, args=None, callopts=None):
        # Feature #4: run `action` through the full pipeline and yield result
        # items, so the `streaming` feature's incremental output is reachable
        # from a generated entity (a normal op call materialises the whole
        # result). `callopts` parameterises the call:
        #   - inbound (download): yield items/chunks (from the streaming
        #     feature when active, else the materialised items);
        #   - outbound (upload): an iterable `body` in callopts is attached to
        #     the request so the transport can stream the payload;
        #   - `ctrl` (pipeline control) and `signal` (cancellation) honoured.
        utility = self._utility

        if callopts is None:
            callopts = {}
        signal = callopts.get("signal")

        ctrl = dict(callopts.get("ctrl") or {})
        ctrl["stream"] = callopts

        ctxmap = {
            "opname": action,
            "ctrl": ctrl,
            "match": self._match,
            "data": self._data,
        }
        if isinstance(args, dict):
            for k, v in args.items():
                ctxmap[k] = v

        ctx = utility.make_context(ctxmap, self._entctx)

        # Outbound: expose the caller's iterable payload so the request builder
        # / transport can stream it as the request body.
        body = callopts.get("body")
        if body is not None:
            ctx.reqdata = dict(ctx.reqdata or {})
            ctx.reqdata["body$"] = body
            ctx.meta["stream_out"] = body

        def aborted():
            if signal is None:
                return False
            if callable(signal):
                return bool(signal())
            return bool(getattr(signal, "aborted", False))

        utility.feature_hook(ctx, "PrePoint")
        point, err = utility.make_point(ctx)
        ctx.out["point"] = point
        if err is not None:
            return

        utility.feature_hook(ctx, "PreSpec")
        spec, err = utility.make_spec(ctx)
        ctx.out["spec"] = spec
        if err is not None:
            return

        utility.feature_hook(ctx, "PreRequest")
        resp, err = utility.make_request(ctx)
        ctx.out["request"] = resp
        if err is not None:
            return

        utility.feature_hook(ctx, "PreResponse")
        resp2, err = utility.make_response(ctx)
        ctx.out["response"] = resp2
        if err is not None:
            return

        utility.feature_hook(ctx, "PreResult")
        result, err = utility.make_result(ctx)
        ctx.out["result"] = result
        if err is not None:
            return

        utility.feature_hook(ctx, "PreDone")

        result = ctx.result

        # Inbound: prefer the streaming feature's incremental generator; else
        # fall back to the materialised items so stream always yields.
        stream_fn = getattr(result, "stream", None) if result is not None else None
        if callable(stream_fn):
            for item in stream_fn():
                if aborted():
                    return
                yield item
        else:
            data = utility.done(ctx)
            if isinstance(data, list):
                items = data
            elif data is None:
                items = []
            else:
                items = [data]
            for item in items:
                if aborted():
                    return
                yield item

    

    
    def list(self, reqmatch=None, ctrl=None) -> list[GetPostalCodesByCity]:
        utility = self._utility
        # reqmatch is optional: an omitted match lists all records. Treat None
        # as an empty match so client.GetPostalCodesByCity().list() works with no args.
        if reqmatch is None:
            reqmatch = {}
        ctx = utility.make_context({
            "opname": "list",
            "ctrl": ctrl,
            "match": self._match,
            "data": self._data,
            "reqmatch": reqmatch,
        }, self._entctx)

        def post_done():
            if ctx.result is not None:
                if ctx.result.resmatch is not None:
                    self._match = ctx.result.resmatch

        return self._run_op(ctx, post_done)



    

    

    

    def _run_op(self, ctx, post_done):
        utility = self._utility

        try:
            utility.feature_hook(ctx, "PrePoint")

            point, err = utility.make_point(ctx)
            ctx.out["point"] = point
            if err is not None:
                return utility.make_error(ctx, err)

            utility.feature_hook(ctx, "PreSpec")

            spec, err = utility.make_spec(ctx)
            ctx.out["spec"] = spec
            if err is not None:
                return utility.make_error(ctx, err)

            utility.feature_hook(ctx, "PreRequest")

            resp, err = utility.make_request(ctx)
            ctx.out["request"] = resp
            if err is not None:
                return utility.make_error(ctx, err)

            utility.feature_hook(ctx, "PreResponse")

            resp2, err = utility.make_response(ctx)
            ctx.out["response"] = resp2
            if err is not None:
                return utility.make_error(ctx, err)

            utility.feature_hook(ctx, "PreResult")

            result, err = utility.make_result(ctx)
            ctx.out["result"] = result
            if err is not None:
                return utility.make_error(ctx, err)

            utility.feature_hook(ctx, "PreDone")

            post_done()

            return utility.done(ctx)

        except Exception:
            utility.feature_hook(ctx, "PreUnexpected")

            raise

File: rb/entity/get_postal_codes_by_city_entity.rb
# ZippopotamusZipCode SDK GetPostalCodesByCity entity

require_relative '../utility/struct/voxgig_struct'
require_relative '../core/helpers'

class GetPostalCodesByCityEntity
  def initialize(client, entopts = nil)
    entopts ||= {}
    if entopts["active"].nil?
      entopts["active"] = true
    elsif entopts["active"] == false
      # keep false
    else
      entopts["active"] = true
    end

    @_name = "get_postal_codes_by_city"
    @_client = client
    @_utility = client.get_utility
    @_entopts = entopts
    @_data = {}
    @_match = {}

    @_entctx = @_utility.make_context.call({
      "entity" => self,
      "entopts" => entopts,
    }, client.get_root_ctx)

    @_utility.feature_hook.call(@_entctx, "PostConstructEntity")
  end

  def get_name
    @_name
  end

  def make
    opts = @_entopts.dup
    GetPostalCodesByCityEntity.new(@_client, opts)
  end

  def data_set(args)
    if args
      @_data = ZippopotamusZipCodeHelpers.to_map(VoxgigStruct.clone(args)) || {}
      @_utility.feature_hook.call(@_entctx, "SetData")
    end
  end

  # @return [GetPostalCodesByCity, Hash] the current GetPostalCodesByCity data
  def data_get
    @_utility.feature_hook.call(@_entctx, "GetData")
    VoxgigStruct.clone(@_data)
  end

  def match_set(args)
    if args
      @_match = ZippopotamusZipCodeHelpers.to_map(VoxgigStruct.clone(args)) || {}
      @_utility.feature_hook.call(@_entctx, "SetMatch")
    end
  end

  # @return [Hash] the current match filter (any subset of GetPostalCodesByCity fields)
  def match_get
    @_utility.feature_hook.call(@_entctx, "GetMatch")
    VoxgigStruct.clone(@_match)
  end

  

  
  # List GetPostalCodesByCity items matching the given filter.
  #
  # @param reqmatch [GetPostalCodesByCityListMatch, Hash, nil] match filter (any subset of
  #   GetPostalCodesByCity fields); defaults to nil, treated as an empty match that lists all.
  # @param ctrl [Object, nil] optional per-call control
  # @return [Array<GetPostalCodesByCity>, Array] the matching GetPostalCodesByCity items; raises ZippopotamusZipCodeError on failure
  def list(reqmatch = nil, ctrl = nil)
    utility = @_utility
    ctx = utility.make_context.call({
      "opname" => "list",
      "ctrl" => ctrl,
      "match" => @_match,
      "data" => @_data,
      "reqmatch" => reqmatch,
    }, @_entctx)

    records = _run_op(ctx) do
      if ctx.result
        @_match = ctx.result.resmatch if ctx.result.resmatch
      end
    end

    # list yields the BARE Array of records — each an accessible Hash — so
    # callers can index item["id"] directly, matching py/lua/go. make_result
    # wraps each entry as an Entity instance for internal use; unwrap those
    # back to their bare record Hashes here (load/create/etc. are unaffected).
    if records.is_a?(Array)
      records = records.map do |item|
        item.respond_to?(:data_get) ? item.data_get : item
      end
    end

    records
  end



  

  

  

  private

  def _run_op(ctx, &post_done)
    utility = @_utility

    # #PrePoint-Hook

    point, err = utility.make_point.call(ctx)
    ctx.out["point"] = point
    return utility.make_error.call(ctx, err) if err

    # #PreSpec-Hook

    spec, err = utility.make_spec.call(ctx)
    ctx.out["spec"] = spec
    return utility.make_error.call(ctx, err) if err

    # #PreRequest-Hook

    resp, err = utility.make_request.call(ctx)
    ctx.out["request"] = resp
    return utility.make_error.call(ctx, err) if err

    # #PreResponse-Hook

    resp2, err = utility.make_response.call(ctx)
    ctx.out["response"] = resp2
    return utility.make_error.call(ctx, err) if err

    # #PreResult-Hook

    result, err = utility.make_result.call(ctx)
    ctx.out["result"] = result
    return utility.make_error.call(ctx, err) if err

    # #PreDone-Hook

    post_done.call

    utility.done.call(ctx)
  end
end
